### DISCUSSION TITLE
[Clickhouse] Ibis significantly slower than clickhouse native client for python

### DISCUSSION THREAD

**Discussion by rongxin-yu** [ibis-project/ibis#6281] [Q&A] [answered]
Hi,

I am testing using ibis against clickhouse and found that ibis seems to run significantly slower than native clickhouse client. I wonder if someone could help explain the reason and advise on how to solve it.

My experiment is as follows:
1. I write IBIS expression and run .execute() method (still running after 5 minutes)
2. I convert the ibis expression to sql and run the sql with native clickhouse client (10 seconds)

I currently plan to go with the second way, only using ibis as a SQL generator, but it's not as convenient as calling the .execute() method directly. 

Any advice is greatly appreciated. Thanks!

**Comment by gforsyth** [comment]
Hey @rongxin-yu -- do you have an example to reproduce this behavior?  That would help us track down what sounds like a bug.  

We expect Ibis to be a touch slower than using any backend directly, but on the order of seconds or fractions of seconds.

**Reply by rongxin-yu** [reply]
Hi @gforsyth, thank you for getting back to me. 

To reproduce, you can follow the following steps.

1. Generate a time series data as follows

- 3000 ids / minute * 400 minutes / day * 250 days / year * 10 years (20130101-20230520)~ 3 billion rows
- have date, time, id, feature 1 (4 columns)

2. suppose you name the table as test, you can reproduce my query as follows

`client = ibis.clickhouse.connect()`
`t = client.table('test')`
`t = t.filter(t.date >='2023-05-01')`
`w = ibis.trailing_window(group_by=['id', 'time'], order_by='date', preceding=30)`
`q=t.mutate(avg=t.feature_1.mean().over(w))`
`q.execute()`

I tried to run it in 3 ways:
a. ibis with duckdb: ~10s
b. convert to sql and run with clickhouse native python client: ~10s
c. ibis with clickhouse: slow

Since the query returns lots of rows, I wonder if it might have anything to do with the protocal, like ADBC vs ODBC/JDBC

**Reply by cpcloud** [reply]
@rongxin-yu A couple of questions that would help us diagnose the issue:

- What version of ibis are you using?
- What is the `CREATE TABLE` statement you used to create `test`?
- What version of ClickHouse server are you running?

Using `master` I was able to produce a table similar to what you describe and run your query with minimal difference between native ibis versus running the SQL string directly.

[Here's a gist](https://gist.github.com/cpcloud/af29ff80118c2377668d271c2b397867) with the test code I used.

Example run:

```
❯ python ~/genslowdata.py --create
100%|████████████████████████████████████████████████████████████| 3250800000/3250800000 [08:49<00:00, 6134431.29it/s]
3250800000

Ibis: 6.07s
SQL: 5.78s
Ibis v. SQL: 0.30s
```

**Reply by yulongkelly** [reply]
Hi @cpcloud, thanks for the followup
Regarding your questions:
-  the clickhouse server version is **23.4.2.11**  and the ibis version is **5.1.0**
- `CREATE TABLE test_table ENGINE = ReplacingMergeTree primary key (date, time, id) order by (date, time, id) settings disk='s3_cache', allow_nullable_key=1 AS SELECT * from file('path/to/*.parquet', 'Parquet');`

**Reply by cpcloud** [reply]
Ok, I edited the above linked gist to use `ReplacingMergeTree` and a similar size table and `CREATE TABLE` and I still don't see a large difference:

First create the data (this will also perform a comparison):

```
❯ python ~/genslowdata.py --table-name test_rmergetree --engine rmergetree --create
```

Test run:

```
❯ python ~/genslowdata.py -t test_rmergetree
table: test_rmergetree
Ibis: 2.73s
SQL: 2.09s
Ibis v. SQL: 0.64s
```

**Reply by cpcloud** [reply]
Can you show the schema of `test_table` with the specific column types?

**Reply by cpcloud** [reply]
I am now trying against a table that has some percentage of nulls in each column.

**Reply by cpcloud** [reply]
Using a table with nulls in every column results in still an order of ~seconds difference:

```
table: test_rmergetree
Ibis: 16.09s
SQL: 13.22s
Ibis v. SQL: 2.87s
```

**Reply by cpcloud** [reply]
I tried against the 5.1.0 release and indeed there's a huge performance hit.

Given that we've moved away from `clickhouse_driver` to `clickhouse_connect` and the problem isn't reproducible on `master`, it seems like this isn't an issue anymore?

Can you give `master` a try on your use case(s)?

**Reply by rongxin-yu** [reply]
Hi @cpcloud 

Thanks for the help.

How should I try the master? I tried the following command, but it didn't seem to solve the problem. Did I miss anything? 

`pip install git+https://github.com/ibis-project/ibis.git@master  `

**Reply by cpcloud** [reply]
Can you try the following?

```
pip install git+https://github.com/ibis-project/ibis.git#egg=ibis-framework[clickhouse]
```

**Reply by rongxin-yu** [reply]
@cpcloud 

I tried to run the following command but still ran into this issue. Is there any way I can check whether I have actually checked out the master branch?

`pip install 'git+https://github.com/ibis-project/ibis.git#egg=ibis-framework[clickhouse]' `

`DEPRECATION: git+https://github.com/ibis-project/ibis.git#egg=ibis-framework[clickhouse] contains an egg fragment with a non-PEP 508 name pip 25.0 will enforce this behaviour change. A possible replacement is to use the req @ url syntax, and remove the egg fragment. Discussion can be found at https://github.com/pypa/pip/issues/11617`

`
Installing collected packages: zstandard, clickhouse-connect
  Attempting uninstall: clickhouse-connect
    Found existing installation: clickhouse-connect 0.3.6
    Uninstalling clickhouse-connect-0.3.6:
      Successfully uninstalled clickhouse-connect-0.3.6
Successfully installed clickhouse-connect-0.5.25 zstandard-0.21.0
`

**Reply by cpcloud** [reply]
@rongxin-yu Thanks for giving that a try.

Were you able to execute that install command in a fresh environment to rule out any pip installation issues?

**Reply by rongxin-yu** [reply]
Hi @cpcloud, 

I just tried to start refresh. 
With version `5.1.0-post.411+53ed88e92`, the issue is gone.

Thanks for the help.

**Reply by cpcloud** [reply, marked as answer]
Awesome!

For future people who may observe this issue, the solution is: *upgrade ibis* :)

This change will be in the next release (6.0.0) which is slated for the end of June.
